### PR TITLE
Implement `fn insert`

### PR DESCRIPTION
**Pull request by Jujumba**
Partially addresses #92

### CODE REVIEW

**Review by pczarn — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 1]
```diff
@@ -1619,6 +1619,30 @@ impl<B: BitBlock> BitVec<B> {
     pub fn shrink_to_fit(&mut self) {
         self.storage.shrink_to_fit();
     }
+
+    /// Inserts a given bit at index `at`
+    ///
+    /// # Examples
+    ///```
+    /// use bit_vec::BitVec;
+    /// let mut b = BitVec::new();
```
Comment (pczarn) [inline src/lib.rs:1628]: Could have a newline after the import, because imports are kind of not program code and no point adding space between program code parts if you have no space from imports

File: src/lib.rs [hunk 2]
```diff
@@ -1619,6 +1619,30 @@ impl<B: BitBlock> BitVec<B> {
...
+    /// b.push(false);
+    /// b.insert(1, true);
+    ///
+    /// assert_eq!(b.len(), 3);
+    /// assert!(b.eq_vec(&[true, true, false]));
+    ///```
+    pub fn insert(&mut self, at: usize, bit: bool) {
+        assert!(
+            at <= self.nbits,
+            "insertion index (is {at}) should be <= nbits (is {len})",
```
[the remark this answers is not in the document]
Comment (pczarn) [reply src/lib.rs:1639]: Could be more like `std::vec`'s docs with the following

> Time complexity

> Takes O(Bit[Vec::len](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.len)) time. All items after the insertion index must be shifted to the right. In the worst case, all elements are shifted when the insertion index is 0.

File: src/lib.rs [hunk 3]
```diff
@@ -1619,6 +1619,30 @@ impl<B: BitBlock> BitVec<B> {
...
+            len = self.nbits
+        );
+        let mut rem = self.split_off(at);
```
[the remark this answers is not in the document]
Comment (Jujumba) [reply src/lib.rs:1642]: Okay, I will write a manual implementation :)

**Comment by Jujumba** [comment]
I've partially rewritten that (passes tests which I've written, though more edge cases __must__ be covered)

I want you to look at the code so far: do it smell?

@pczarn

**Comment by pczarn** [comment]
@Jujumba 

This is the most optimal way to do it. I left the second part to finish, but the first part may work though it's untested.

```rust
    pub fn insert(&mut self, at: usize, bit: bool) {
        assert!(
            at <= self.nbits,
            "insertion index (is {at}) should be <= nbits (is {nbits})",
            nbits = self.nbits
        );

        let last_block_bits = self.nbits % B::bits();
        let block_at = at / B::bits(); // number of blocks
        let bit_at = at % B::bits(); // index within a block

        if last_block_bits == 0 {
            self.storage.push(B::zero());
        }

        let mut carry = self.storage[block_at] >> (B::bits() - 1);
        let lsbits_mask = (B::one() << bit_at) - 1;
        let set_bit = if bit { B::one() } else { B::zero() } << bit_at;
        self.storage[block_at] = (self.storage[block_at] & lsbits_mask) | ((self.storage[block_at] & !lsbits_mask) << 1) | set_bit;

        for block_ref in &mut self.storage[block_at + 1 ..] {
            // carry = ...
            // *block_ref = *block_ref << 1 | ...
        }
    }
```

**Comment by pczarn** [comment]
The tests you need are edge cases for bit 0, 31, 32, 64, bit of the last block, last bit of the last block, as well as various lengths of the bit-vec. Could do a test for a panic when index is out of bounds.

**Review by pczarn — CHANGES_REQUESTED** [review]
As in previous comments

**Comment by Jujumba** [comment]
> @Jujumba
> 
> This is the most optimal way to do it. I left the second part to finish, but the first part may work though it's untested.
> 
> ```rust
>     pub fn insert(&mut self, at: usize, bit: bool) {
>         assert!(
>             at <= self.nbits,
>             "insertion index (is {at}) should be <= nbits (is {nbits})",
>             nbits = self.nbits
>         );
> 
>         let last_block_bits = self.nbits % B::bits();
>         let block_at = at / B::bits(); // number of blocks
>         let bit_at = at % B::bits(); // index within a block
> 
>         if last_block_bits == 0 {
>             self.storage.push(B::zero());
>         }
> 
>         let mut carry = self.storage[block_at] >> (B::bits() - 1);
>         let lsbits_mask = (B::one() << bit_at) - 1;
>         let set_bit = if bit { B::one() } else { B::zero() } << bit_at;
>         self.storage[block_at] = (self.storage[block_at] & lsbits_mask) | ((self.storage[block_at] & !lsbits_mask) << 1) | set_bit;
> 
>         for block_ref in &mut self.storage[block_at + 1 ..] {
>             // carry = ...
>             // *block_ref = *block_ref << 1 | ...
>         }
>     }
> ```

Thanks for you clues (=^ ◡ ^=)

**Comment by Jujumba** [comment]
I think we are probably done, though I suspect not all edge-cases are covered

File: src/lib.rs [hunk 4]
```diff
@@ -1619,6 +1619,61 @@ impl<B: BitBlock> BitVec<B> {
...
+    /// use bit_vec::BitVec;
+    ///
+    /// let mut b = BitVec::new();
+    ///
+    /// b.push(true);
+    /// b.push(true);
+    /// b.insert(1, false);
+    ///
+    /// assert_eq!(b.len(), 3);
+    /// assert!(b.eq_vec(&[true, false, true]));
```
Comment (pczarn) [inline src/lib.rs:1639]: No need to test for len if you test for eq_vec.

File: src/lib.rs [hunk 5]
```diff
@@ -1619,6 +1619,61 @@ impl<B: BitBlock> BitVec<B> {
...
+    ///```
+    ///
+    /// # Time complexity                                                                                                                                                         
+    /// Takes O([`BitVec::len`]) time. All items after the insertion index must be
```
Comment (pczarn) [inline src/lib.rs:1643]: Please add a fragment specifier below

[\`BitVec::len\`]: Self::len

File: src/lib.rs [hunk 6]
```diff
@@ -1619,6 +1619,61 @@ impl<B: BitBlock> BitVec<B> {
...
+
+        let last_block_bits = self.nbits % B::bits();
+        let block_at = at / B::bits(); // needed block
+        let bit_at = at % B::bits(); // index within the block
+
+        if last_block_bits == 0 {
+            self.storage.push(B::zero());
+        }
+
+        self.nbits += 1;
```
Comment (pczarn) [inline src/lib.rs:1662]: :+1:

File: src/lib.rs [hunk 7]
```diff
@@ -1619,6 +1619,61 @@ impl<B: BitBlock> BitVec<B> {
     pub fn shrink_to_fit(&mut self) {
         self.storage.shrink_to_fit();
     }
+
+    /// Inserts a given bit at index `at`, shifting all bits after by one
+    ///
+    /// # Panics
+    /// Panics if `at > nbits`
```
Comment (pczarn) [inline src/lib.rs:1626]: Could be more clear, saying Panics if the `at` argument is out of bounds for `BitVec`'s length.

File: src/lib.rs [hunk 8]
```diff
@@ -1619,6 +1619,61 @@ impl<B: BitBlock> BitVec<B> {
...
+        let lsbits_mask = (B::one() << bit_at) - B::one();
+        let set_bit = if bit { B::one() } else { B::zero() } << bit_at;
+        self.storage[block_at] = (self.storage[block_at] & lsbits_mask)
+            | ((self.storage[block_at] & !lsbits_mask) << 1)
+            | set_bit;
+
+        for block_ref in &mut self.storage[block_at + 1..] {
+            let curr_carry = *block_ref >> (B::bits() - 1);
+            *block_ref = *block_ref << 1 | carry;
+            carry = curr_carry;
```
Comment (pczarn) [inline src/lib.rs:1674]: :+1:

File: src/lib.rs [hunk 9]
```diff
@@ -3053,4 +3108,77 @@ mod tests {
...
+        assert_eq!(v.storage().len(), 2);
+    }
+
+    #[test]
+    fn test_insert_at_block_boundaries_1() {
+        let mut v = BitVec::from_elem(64, false);
+
+        assert_eq!(v.storage().len(), 2);
+
+        v.insert(63, true);
```
Comment (pczarn) [inline src/lib.rs:3170]: :+1:

File: src/lib.rs [hunk 10]
```diff
@@ -1619,6 +1619,61 @@ impl<B: BitBlock> BitVec<B> {
...
+    ///
+    /// assert_eq!(b.len(), 3);
+    /// assert!(b.eq_vec(&[true, false, true]));
+    ///```
+    ///
+    /// # Time complexity                                                                                                                                                         
+    /// Takes O([`BitVec::len`]) time. All items after the insertion index must be
+    /// shifted to the right. In the worst case, all elements are shifted when
+    /// the insertion index is 0.
+
```
Comment (pczarn) [inline src/lib.rs:1646]: Nit: Unncesessary newline.

**Review by pczarn — CHANGES_REQUESTED** [review]
Looks close to good.

File: src/lib.rs [hunk 11]
```diff
@@ -3053,4 +3108,77 @@ mod tests {
...
+            false, false, false, false, false, false, false, false, false, false, false, false,
+            false, false, false, false, false, false, false, false, false, false, false, false,
+            false, false, false, false, false, false, false, true, false
+        ]));
+
+        assert_eq!(v.storage().len(), 2);
+    }
+
+    #[test]
+    fn test_insert_at_block_boundaries_1() {
```
Comment (pczarn) [inline src/lib.rs:3165]: We really could benefit a lot in our project from the `test_case` lib dev dependency for tests.

File: src/lib.rs (hunk 4 above)
Comment (Jujumba) [reply src/lib.rs:1639]: Yeah, I just wanted it to be more explicit

File: src/lib.rs (hunk 5 above)
Comment (pczarn) [reply src/lib.rs:1642]: Then I'm not sure why one of our contributors was adding fragment specifiers. Ah, it's because he was changing all stuff like `BitVec::len` to just `len`.

Comment (pczarn) [reply src/lib.rs:1642]: @Jujumba you may change to just `len`

**Review by pczarn — CHANGES_REQUESTED** [review]
As in comments above; it's almost ready; I don't see any more stuff except the nit for doc comment

**Review by pczarn — APPROVED** [review]
@Jujumba  Merging! Thanks!!! This is the right kind of work for me since I am now recovering from being sick, so I'd much prefer to review than get tired with coding or do nothing at all.

This is a pretty fundamental and popular library, so I'm asking you for cleaner code than would be needed in some other kinds of code.

You know, I took a look and ferrilab's bitvec is doing a push of one bit at the end, then calling a function to rotate slice by one bit, well we do not have bitslice implemented to do that, besides this code here is better because calling rotate in such code is a bit confusing and weird for the readers. I wouldn't want to copy their code anyway as their licensing is a bit different. So this code is okay. In the future it would be best to add `BitSlice` but this is waiting on an important Rust DST improvement, and  once that is added we can put your code in `fn shift_right` and run as `bit_slice.shift_right(1)` for `&mut BitSlice`  as  well as write rotate_left/right too like theirs.

**Comment by Jujumba** [comment]
> @Jujumba Merging! Thanks!!! This is the right kind of work for me since I am now recovering from being sick, so I'd much prefer to review than get tired with coding or do nothing at all.
> 
> This is a pretty fundamental and popular library, so I'm asking you for cleaner code than would be needed in some other kinds of code.
> 
> You know, I took a look and ferrilab's bitvec is doing a push of one bit at the end, then calling a function to rotate slice by one bit, well we do not have bitslice implemented to do that, besides this code here is better because calling rotate in such code is a bit confusing and weird for the readers. I wouldn't want to copy their code anyway as their licensing is a bit different. So this code is okay. In the future it would be best to add `BitSlice` but this is waiting on an important Rust DST improvement, and once that is added we can put your code in `fn shift_right` and run as `bit_slice.shift_right(1)` for `&mut BitSlice` as well as write rotate_left/right too like theirs.

Good to hear that

I wish you to get well soon :)
